### DISCUSSION TITLE
Updating request headers according to request body (Custom Authentication)

### DISCUSSION THREAD

**Discussion by byildiz** [aio-libs/aiohttp#6627] [Q&A] [answered]
Hi everyone,

I am trying to implement a type of custom authentication by using `aiohttp` something like the example in [this link](https://docs.python-requests.org/en/latest/user/advanced/#custom-authentication) but I also need request body. Here is an example for [requests](https://docs.python-requests.org/en/latest/):

```py
class CustomAuth(AuthBase):
    def __init__(self, secretkey):
        self.secretkey = secretkey

    def get_hash(self, request):
        if request.body:
            data = request.body.decode('utf-8')
        else:
            data = "{}"

        signature = hmac.new(
        	str.encode(self.secretkey),
        	msg=str.encode(data),
            digestmod=hashlib.sha256
        ).hexdigest().upper()
        return signature

    def __call__(self, request):
        request.headers["CUSTOM-AUTH"] = self.get_hash(request)
        return request
```

I've looked into [tracing](https://docs.aiohttp.org/en/stable/tracing_reference.html) and [BasicAuth](https://docs.aiohttp.org/en/stable/client_reference.html#basicauth) but they are useless in my situation. On `on_request_start` request body is not ready, on `on_request_chunk_sent` headers have already been sent. A solution like `BasicAuth` don't have access the request data at all.

Do you have any idea?

Thanks in advance.

**Comment by BahmanBinary** [comment]
i have this problem too.
can anyone help?

**Comment by BahmanBinary** [comment]
hi @byildiz 
is your problem solved?

**Reply by byildiz** [reply]
Hi,
I found a workaround: I create request body myself using json dump and give it to aiohttp. So I can do what ever I want with the body.

**Comment by byildiz** [comment, marked as answer]
I think I should write my workaround as an answer:

I solved my problem by using `json.dumps` since aiohttp internally uses `json.dumps` to convert json data to http body by default. Here it is:
https://github.com/aio-libs/aiohttp/blob/ed196893b6102cedcc98251c0c76d1b9c5103cee/aiohttp/client.py#L205
You can call `json.dumps` on the data that you want to sent and you will get the exactly same body with the body aiohttp will create.

**Reply by BahmanBinary** [reply]
thank u for your help @byildiz
